Chore: Add Support for Automated Publishing

jordanleven opened the pull request:
# Description

This PR adds support for automated publishing to the NPM public registry. In addition, this PR contains the following:

- Untracks distribution files from Git (files are now built as part of the pipeline and published via a GitHub action).
- Consolidates all source code within the `src` directory.
- Adds support for preprocessing Scss.
- Removes unnecessary dependencies.
- Removes non-functioning NPM scripts.
- Security updates for NPM packages.


# Validation Steps

## Prerequisites
1. Apply the supplied [patch](https://github.com/sparkbox/downpour/files/6772165/patch_chore--add-support-for-automated-publishing.zip). This will allow you to both simulate releases and publish locally to Verdaccio.
2. Install [Verdaccio](https://verdaccio.org). This will allow you to publish the artifact locally without hitting the public registry.

## Validation Steps
1. Checkout the validation branch (`chore--add-support-for-automated-publishing--validation`).
- [x] Confirm only [additional commit in validation branch has a commit subject of "DROP"](https://github.com/sparkbox/downpour/commits/chore--add-support-for-automated-publishing--validation).
1. Install all dependencies (`npm i`).
1. Run linting task (`npm run lint`).
- [x] Confirm no linting violations appear.
3. Run `npm run release`.
1. Confirm the new release number by hitting "y".
- [x] Confirm you receive the following output (note that even though it will say, "Package [patch_chore--add-support-for-automated-publishing.zip](https://github.com/sparkbox/downpour/files/6772161/patch_chore--add-support-for-automated-publishing.zip) updated and pushed to the remote." it has not).
```sh

✔ bumping version in package.json from 0.0.12 to 0.0.13
✔ bumping version in package-lock.json from 0.0.12 to 0.0.13
✔ created CHANGELOG.md
✔ outputting changes to CHANGELOG.md

---
### [0.0.13](https://github.com/sparkbox/downpour/compare/v0.0.12...v0.0.13) (2021-07-06)
---
```
5. Start Verdaccio (`verdaccio`)
1. Run `npm publish --registry http://localhost:4873` to build and publish the artifact to Verdaccio.
1. In your project, run `npm rm @sparkbox/downpour && npm i @sparkbox-local/downpour`.
1. In your project, update the import from `@sparkbox/downpour` to `@sparkbox-local/downpour`.
- [x] Confirm the project runs as expected.

# Post-Merge Reminders
- [x] Delete `chore--add-support-for-automated-publishing--validation` from GitHub.

jordanleven committed:
refactor: Update build process

Build process
- Previously, both source code and transpiled distribution code were
  tracked in Git. In our new process, distribution files are not tracked
  and are built using the new build script. In the future, distribution
  files will be generated by our CI and shouldn't be tracked in Git. By
  building files in CI, we avoid issues where a user pushes changes that
  do not include the most up-to-date build.
- Dist files are now transpiled via Rollup instead of Webpack.

Update README.md
- Fix import path in docs.
- Update import syntax.
- Use footer links in markdown.

Add support for Sass
- Although our styling is currently CSS, this adds support for Sass (if
  contributors wanted to use Sass features in the future).

Clean up package.json
- `docs` don't exist in this project, so they should be listed as a
  directory.
- Remove dependencies that are listed in the package.json but are not
  actively being used.
- Remove scripts in package.json that are not being used.

reedspool commented on package.json:
```diff
@@ -34,33 +29,22 @@
+    "eslint": "^7.29.0",
+    "eslint-config-airbnb": "^18.2.1",
+    "node-sass": "^6.0.1",
+    "rollup": "^2.52.7",
```
Node-sass is deprecated. `npm install sass` has the same API. I presume it will work in rollup too.

jordanleven replied:
Good call - I'll update this.

reedspool commented on package.json:
```diff
@@ -5,21 +5,16 @@
   "engines": {
     "node": ">=6.9.1"
```
This version of node is long since departed. You might as well bump this to whatever the lowest version you're able to test your changes on.

jordanleven replied:
Oh yeah, I haven't gotten around to that (this PR is still a draft PR). In fact, you might notice that none of the automated publishing works yet!

reedspool commented on src/js/downpour.js:
```diff
@@ -1,13 +1,13 @@
-const ifProd = require('@sparkbox/if-prod');
-const shelljs = require('shelljs');
-const _ = require('lodash');
+import '../css/drizzle.scss';
```
I feel this magic line needs a comment. Since there are no tests asserting the correct output of the CSS file, it would be too easy for someone to delete this weird line. That hapless dev might not notice until downstream npm users complained at errors.

jordanleven committed:
chore: Add support for new GitHub Actions

Add support for automating release
- Adds new actions for publishing to NPM when new version of the
  package are cut.

Add support for CI action
- New CI action will run to make sure application code is free of
  linting errors, non-standard commit subjects.

reedspool commented:
I successfully performed the validation steps and confirmed the NPM installation works as expected in a project which uses this.

robtarr commented on package.json:
```diff
@@ -3,7 +3,7 @@
   "description": "Sparkbox implementation of drizzle-builder",
   "engines": {
-    "node": ">=12.0.0"
+    "node": ">=14.0.0"
```
Do we also want to change the GitHub CI action to use 14?

jordanleven replied:
Ah, great call! All updated in 5ebc3295e6105b3c5bbbfbdbd982514f31b7656d.
